closed/src/java.base/share/native/libjncrypto/NativeCrypto.c

```diff
@@ -474,12 +576,260 @@ JNIEXPORT jlong JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_loadCrypto
+#if defined(_AIX)
+        int flags = RTLD_NOW;
+        if (NULL != strrchr(libName, '(')) {
+            flags |= RTLD_MEMBER;
+        }
+        result = dlopen(libName, flags);
+#elif defined(_WIN32) /* defined(_AIX) */
+        result = LoadLibrary(libName);
+#else /* defined(_WIN32) */
+        result = dlopen(libName, RTLD_NOW);
+#endif /* defined(_AIX) */
+    }
+    return result;
+}
+
+/* Look for a crypto library in java.home or the system.
+ * NULL is returned when an appropriate crypto library
+ * cannot be found.
+ */
+static void *
+find_crypto_library(jboolean traceEnabled, const char *chomepath)
+{
+    void *result = NULL;
+    void *prevResult = NULL;
+    size_t numOfLibs = 0;
+    size_t i = 0;
+    long tempVersion = 0;
+    long previousVersion = 0;
+    size_t num_of_generic = 0;
+
+#if defined(_AIX)
+    num_of_generic = 4;
+#elif defined(__linux__) /* defined(_AIX) */
+    num_of_generic = 1;
+#endif /* defined(_AIX) */
+
+    /* Library names for OpenSSL 1.1.1, 1.1.0 and symbolic links:
+     * It is important to preserve the order!!!
+     */
+    static const char * const libNames[] = {
+#if defined(_AIX)
+        "libcrypto.a(libcrypto64.so)",      /* general symlink library name from archive file */
+        "libcrypto64.so",                   /* general symlink library name */
+        "libcrypto.a(libcrypto.so)",        /* general symlink library name from archive file */
+        "libcrypto.so",                     /* general symlink library name */
+        "libcrypto.a(libcrypto64.so.3)",    /* 3.x library name from archive file */
+        "libcrypto64.so.3",                 /* 3.x library name */
+        "libcrypto.a(libcrypto.so.3)",      /* 3.x library name from archive file */
+        "libcrypto.so.3",                   /* 3.x library name */
+        "libcrypto.a(libcrypto64.so.1.1)",  /* 1.1.x library name from archive file */
+        "libcrypto.so.1.1",                 /* 1.1.x library name */
+        "libcrypto.so.1.0.0",               /* 1.0.x library name */
+#elif defined(__APPLE__) /* defined(_AIX) */
+        "libcrypto.3.dylib",                /* 3.x library name */
+        "libcrypto.1.1.dylib",              /* 1.1.x library name */
+        "libcrypto.1.0.0.dylib",            /* 1.0.x library name */
+#elif defined(_WIN32) /* defined(__APPLE__) */
+        "libcrypto-3-x64.dll",              /* 3.x library name */
+        "libcrypto-1_1-x64.dll",            /* 1.1.x library name */
+        "libeay32.dll",                     /* old library name */
+#else /* defined(_WIN32) */
+        "libcrypto.so",                     /* general symlink library name */
+        "libcrypto.so.3",                   /* 3.x library name */
+        "libcrypto.so.1.1",                 /* 1.1.x library name */
+        "libcrypto.so.1.0.0",               /* 1.0.x library name */
+        "libcrypto.so.10",                  /* old library name */
+#endif /* defined(_AIX) */
```

Comment:
Please explain why the search order is changed. For example, on AIX, the symbolic link is sought first instead of last and there are four entries that are expected to by symbolic links. On macOS, perhaps the symlink entry should stay despite the discussion https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/884#discussion_r1852426196.

Reply:
I've enhanced the comment to explain the reasoning behind the need for this order and I've re-added the `macOS` general symlink.